File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка строк/main.c
#include <stdio.h>
#include <string.h>

#define MAX_CHAR 256

// Функция сортировки подсчётом по символам в строках
void counting_sort(char* arr[], int n, int char_index) {
    int count[MAX_CHAR] = {0};
    char* output[n];

    // Подсчёт количества каждого символа в текущем разряде
    for (int i = 0; i < n; i++) {
        int char_code = arr[i][char_index] ? arr[i][char_index] : 0;
        count[char_code]++;
    }

    // Преобразование count в массив позиций
    for (int i = 1; i < MAX_CHAR; i++) {
        count[i] += count[i - 1];
    }

    // Строим отсортированный массив
    for (int i = n - 1; i >= 0; i--) {
        int char_code = arr[i][char_index] ? arr[i][char_index] : 0;
        output[count[char_code] - 1] = arr[i];
        count[char_code]--;
    }

    // Копирование отсортированных строк в исходный массив
    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

// Основная функция для сортировки строк
void radix_sort_strings(char* arr[], int n) {
    // Находим максимальную длину строки
    int max_len = 0;
    for (int i = 0; i < n; i++) {
        int len = strlen(arr[i]);
        if (len > max_len) {
            max_len = len;
        }
    }

    // Сортируем строки по каждому символу (начиная с последнего)
    for (int char_index = max_len - 1; char_index >= 0; char_index--) {
        counting_sort(arr, n, char_index);
    }
}

// Функция для вывода массива строк
void print_strings(char* arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%s ", arr[i]);
    }
    printf("\n");
}

int main() {
    char* arr[] = {"apple", "banana", "grape", "cherry", "date"};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("До сортировки:\n");
    print_strings(arr, n);

    radix_sort_strings(arr, n);

    printf("После сортировки:\n");
    print_strings(arr, n);

    return 0;
}

File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка строк/main.py
def counting_sort(arr, char_index):
    # Массив для подсчета символов (от 0 до 255 для ASCII)
    count = [0] * 256
    # Массив для результата сортировки
    output = [''] * len(arr)
    
    # Подсчитаем количество каждого символа в заданном разряде (char_index)
    for string in arr:
        count[ord(string[char_index])] += 1
    
    # Преобразуем count в массив позиций
    for i in range(1, 256):
        count[i] += count[i - 1]
    
    # Строим отсортированный результат
    for string in reversed(arr):  # для стабильности сортировки
        output[count[ord(string[char_index])] - 1] = string
        count[ord(string[char_index])] -= 1
    
    return output

def radix_sort_strings(arr):
    if len(arr) == 0:
        return arr
    
    # Определим максимальную длину строки
    max_len = max(len(s) for s in arr)
    
    # Проводим сортировку по каждому символу от старшего к младшему
    for char_index in range(max_len - 1, -1, -1):
        arr = counting_sort(arr, char_index)
    
    return arr

# Тестирование:
strings = ["apple", "banana", "grape", "cherry", "date"]
sorted_strings = radix_sort_strings(strings)

print("Отсортированные строки:", sorted_strings)
